tests/unit/invert/laplace/test_laplace_petsc3damg.cxx

```diff
@@ -0,0 +1,370 @@
+	newF[i] = (f[i] - f[i.xm()])/ coords->dx[i] / sqrt(coords->g_11[i]);
+      } else {
+	newF[i] = 0.5 * (f[i.xm()] + f[i]);
+      }
+    }
+
+    BOUT_FOR(i, f.getMesh()->getRegion3D("RGN_LOWER_Y")) {
+      if (lower_y_neumann) {
+	newF[i] = (f[i.yp()] - f[i])/ coords->dx[i] / sqrt(coords->g_11[i]);
+      } else {
+	newF[i] = 0.5 * (f[i] + f[i.yp()]);
+      }
+    }
+
+    BOUT_FOR(i, f.getMesh()->getRegion3D("RGN_UPPER_Y")) {
+      if (upper_y_neumann) {
+	newF[i] = (f[i] - f[i.ym()])/ coords->dx[i] / sqrt(coords->g_11[i]);
+      } else {
+	newF[i] = 0.5 * (f[i.ym()] + f[i]);
+      }
+    }
+  }
+};
+
+
+class Petsc3dAmgTest : public FakeMeshFixture,
+		       public testing::WithParamInterface<std::tuple<bool, bool,
+								     bool, bool > > {
+public:
+  WithQuietOutput info{output_info}, warn{output_warn}, progress{output_progress}, all{output};
+  Petsc3dAmgTest() : FakeMeshFixture() {
+    PetscErrorPrintf = PetscErrorPrintfNone;
+    int nx = mesh->GlobalNx,
+        ny = mesh->GlobalNy,
+        nz = mesh->GlobalNz;
+    BoutReal x, y, z;
+    static_cast<FakeMesh*>(bout::globals::mesh)->setGridDataSource(new GridFromOptions(Options::getRoot()));
+    bout::globals::mesh->getCoordinates()->geometry();
+    f3.allocate();
+    coef2.allocate();
+    coef3.allocate();
+
+    BOUT_FOR(i, mesh->getRegion2D("RGN_ALL")) {
+      x = i.x()/(BoutReal)nx - 0.5;
+      y = i.y()/(BoutReal)ny - 0.5;
+      coef2[i] = x + y;
+    }
+    BOUT_FOR(i, mesh->getRegion3D("RGN_ALL")) {
+      x = i.x()/(BoutReal)nx - 0.5;
+      y = i.y()/(BoutReal)ny - 0.5;
+      z = i.z()/(BoutReal)nz - 0.5;
+      f3[i] = 1e3*exp(-0.5*sqrt(x*x + y*y + z*z)/sigmasq);
+      coef3[i] = x + y + sin(2*3.14159265358979323846*z);
+    }
+    auto param = GetParam();
+    forward = ForwardOperator(std::get<0>(param), std::get<1>(param),
+			      std::get<2>(param), std::get<3>(param));
+    Options *options = Options::getRoot()->getSection("laplace");
+    (*options)["type"] = "petsc3damg";
+    (*options)["inner_boundary_flags"] = (std::get<0>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["outer_boundary_flags"] = (std::get<1>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["lower_boundary_flags"] = (std::get<2>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["upper_boundary_flags"] = (std::get<3>(param) ? INVERT_AC_GRAD : 0) + INVERT_RHS;
+    (*options)["fourth_order"] = false;
+    (*options)["atol"] = tol/10; // Need to specify smaller than desired tolerance to
+    (*options)["rtol"] = tol/10; // ensure it is satisfied for every element.
+    //    (*options)["dtol"] = 1e5;
+    //    (*options)["maxits"] = 100000;
+    //    (*options)["richardson_damping_factor"] = 1.0;
+    //    (*options)["chebyshev_max"] = 100.0;
+    //    (*options)["chebyshev_min"] = 0.01;
+    //    (*options)["gmres_max_steps"] = 30;
+    solver = std::unique_ptr<LaplacePetsc3dAmg>(static_cast<LaplacePetsc3dAmg*>(Laplacian::create().release()));
```

Comment:
Probably cleaner to just make a `LaplacePetsc3dAmg` directly!

I take it you need to access some methods that aren't in the base class? Is it possible to write the tests without that?

Reply:
No, it is not possible. I need to access to the `getMatrix3D()` method (incidentally, that should really just be called `getMatrix()`) in order to ensure that it behaves correctly as a forward operator.

Would calling the `LaplacePetsc3dAmg` constructor directly be acceptable? I guess I was under the impression that I shouldn't trust doing that and should stick to using the factory, but looking at the code I don't see a good reason to do so.

Reply:
Yup, should be fine. In fact, I think if you make the member declaration literally `LaplacePetsc3dAmg solver{};` it will work, it doesn't even need to be a `unique_ptr`, although that will mean changing `solver->` to `solver.`.

Reply:
This required a little bit of further refactoring so that the solver could be initialised before the body of the constructor, but this wasn't much of a problem.